In [6]:
#imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
%matplotlib inline
import seaborn as sns
import statsmodels.formula.api as smf

## 3.7.10
This question should be answered using the Carseats data set.

(a) Fit a multiple regression model to predict Sales using Price, Urban, and US.

In [11]:
carseats_df = pd.read_csv("../DataSets/Carseats.csv")
mul_reg = smf.ols('Sales ~ Price + Urban + US', carseats_df).fit()
mul_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Wed, 07 Feb 2024   Prob (F-statistic):           2.39e-23
Time:                        15:23:04   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       13.0435      0.651     20.036      0.000      11.764      14.323
Urban[T.Yes]    -0.0219      0.272     -0.081      0.936      -0.556       0.512
US[T.Yes]        1.2006      0.259      4.635      0.000       0.691       1.710
Price           -0.0545      0.005    -10.389      0.000      -0.065      -0.044
==============================================================================
Omnibus:                        0.676   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.713   Jarque-Bera (JB):                0.758
Skew:                           0.093   Prob(JB):                        0.684
Kurtosis:                       2.897   Cond. No.                         628.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

(b) Provide an interpretation of each coefficient in the model. Be careful—some of the variables in the model are qualitative!

The coefficients are $\beta_0 = 13.0435, \beta_1 = -0.0219, \beta_2 = 1.2006, \beta_3 = -0.0545$. $\beta_0$ is the intercept of the model, or the constant term. $\beta_1$ is the value by which $\beta_0$ increases if the carseat was sold in an urban area. $\beta_2$ is the value by which $\beta_0$ increases if the carseat was sold in the US. $\beta_3$ is the value by which $\beta_0$ increases if the price increases.  

(c) Write out the model in equation form, being careful to handle the qualitative variables properly.


Sales is represented as $Y$. 
$Y = 13.0435 + (-0.0219)*$ Urban[Yes] $+ (1.2006)*$US[Yes] $+ (-0.0545)*$ Price

(d) For which of the predictors can you reject the null hypothesis $H_0 : \beta_j = 0$?


We can reject the Urban predictor because the p-value for this predictor is very high (>0.05). 

(e) On the basis of your response to the previous question, fit a smaller model that only uses the predictors for which there is evidence of association with the outcome.


In [12]:
mul_reg_new = smf.ols('Sales ~ Price + US', carseats_df).fit()
mul_reg_new.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Wed, 07 Feb 2024   Prob (F-statistic):           2.66e-24
Time:                        15:24:02   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.0308      0.631     20.652      0.000      11.790      14.271
US[T.Yes]      1.1996      0.258      4.641      0.000       0.692       1.708
Price         -0.0545      0.005    -10.416      0.000      -0.065      -0.044
==============================================================================
Omnibus:                        0.666   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.717   Jarque-Bera (JB):                0.749
Skew:                           0.092   Prob(JB):                        0.688
Kurtosis:                       2.895   Cond. No.                         607.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 4.8.4
When the number of features p is large, there tends to be a deterioration in the performance of KNN and other local approaches that perform prediction using only observations that are near the test observation for which a prediction must be made. This phenomenon is known as the curse of dimensionality, and it ties into the fact that non-parametric approaches often perform poorly when p is large. We will now investigate this curse.

(a) Suppose that we have a set of observations, each with measurements on $p = 1$ feature, $X$. We assume that $X$ is uniformly (evenly) distributed on $[0, 1]$. Associated with each observation is a response value. Suppose that we wish to predict a test observation’s response using only observations that are within 10 % of the range of $X$ closest to that test observation. For instance, in order to predict the response for a test observation with $X = 0.6$, we will use observations in the range $[0.55, 0.65]$. On average, what fraction of the available observations will we use to make the prediction?

We are using 10% of the range of the interval [0,1], therefore the fraction of available observations we will use is 10%. 

(b) Now suppose that we have a set of observations, each with
measurements on $p = 2$ features, $X_1$ and $X_2$. We assume that $(X_1, X_2)$ are uniformly distributed on $[0, 1] \times [0, 1]$. We wish to predict a test observation’s response using only observations that are within 10 % of the range of $X_1$ and within 10 % of the range of $X_2$ closest to that test observation. For instance, in order to predict the response for a test observation with $X_1 = 0.6$ and
$X_2 = 0.35$, we will use observations in the range $[0.55, 0.65]$ for $X_1$ and in the range $[0.3, 0.4]$ for $X_2$. On average, what fraction of the available observations will we use to make the prediction?

We use 10% of the first interval and 10% of the second interval, therefore in total we have 10% * 10%  = 1%. 

(c) Now suppose that we have a set of observations on $p = 100$ features. Again the observations are uniformly distributed on each feature, and again each feature ranges in value from 0 to 1. We wish to predict a test observation’s response using observations within the 10 % of each feature’s range that is closest to that test observation. What fraction of the available observations will we use to make the prediction?

Using the same method from (b), we have that (0.1)^100 is the fraction of the available observations to be used.

(d) Using your answers to parts (a)–(c), argue that a drawback of KNN when $p$ is large is that there are very few training observations “near” any given test observation.

When p is large we see that very little of the total set of observation is near any given test observation. As dimensions get larger (so p value gets larger) the number of of nearby neighbors goes to 0, therefore at a high enough dimension there are no neighbors in high dimensions. 

## 4.8.6

Suppose we collect data for a group of students in a statistics class with variables $X_1 =$ hours studied, $X_2 =$ undergrad GPA, and $Y =$ receive an A. We fit a logistic regression and produce estimated coefficient, $\hat{\beta_0} = -6, \hat{\beta_1} = 0.05, \hat{\beta_2} = 1$.

(a) Estimate the probability that a student who studies for 40 h and has an undergrad GPA of 3.5 gets an A in the class.

In [15]:
model = -6 + 0.05*40 + 1*3.5
y = np.exp(model) / (1+np.exp(model))
y

0.37754066879814546

(b) How many hours would the student in part (a) need to study to have a 50 % chance of getting an A in the class?

In [16]:
#we have that 0.5 = e^model/(1+e^model)
#so 0.5 + 0.5e^model = e^model
#0.5e^model = 0.5
#e^model = 0.5
#model = 0
#so we want to find values of hours and GPA such that 0.05*hours + GPA = 6
#A is a 3.5
#0.05*h = 6 - 3.5
#h = (6-3.5)/0.05
hours = (6-3.5)/0.05
hours

50.0